larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,503 @@
+{
+
+    LArHitWidthHelper::ClusterToParametersMapStore* pClusterToParametersMapStore = LArHitWidthHelper::ClusterToParametersMapStore::Instance();
+    LArHitWidthHelper::ClusterToParametersMap* pClusterToParametersMap = pClusterToParametersMapStore->GetMap();
+
+    if(pClusterToParametersMap->empty())
+        throw StatusCodeException(STATUS_CODE_NOT_INITIALIZED);
+
+    // ATTN this method assumes that clusters have been sorted by extremal x position (low x -> high x)
+    for (ClusterVector::const_iterator iterCurrentCluster = clusterVector.begin(); iterCurrentCluster != clusterVector.end(); ++iterCurrentCluster)
+    {
+        const Cluster *const pCurrentCluster = *iterCurrentCluster;
+        const LArHitWidthHelper::ClusterToParametersMap::const_iterator currentParametersIter = pClusterToParametersMap->find(pCurrentCluster);
+
+        if(currentParametersIter == pClusterToParametersMap->end())
+            throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+        for (ClusterVector::const_iterator iterTestCluster = iterCurrentCluster; iterTestCluster != clusterVector.end(); ++iterTestCluster)
+        {
+            if (iterCurrentCluster == iterTestCluster)
+                continue;
+
+	    const Cluster *const pTestCluster = *iterTestCluster;
+            const LArHitWidthHelper::ClusterToParametersMap::const_iterator testParametersIter = pClusterToParametersMap->find(pTestCluster);
+
+	    if(testParametersIter == pClusterToParametersMap->end())
+                throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+            if (!this->AreClustersAssociated(currentParametersIter->second, testParametersIter->second))
+	        continue;
+
+            clusterAssociationMap[pCurrentCluster].m_forwardAssociations.insert(pTestCluster);
+            clusterAssociationMap[pTestCluster].m_backwardAssociations.insert(pCurrentCluster);
+        }
+    }
+
+    // remove all 'shortcut' routes
+    this->CleanupClusterAssociations(clusterVector, clusterAssociationMap);
+
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void HitWidthClusterMergingAlgorithm::CleanupClusterAssociations(const ClusterVector &clusterVector, ClusterAssociationMap &clusterAssociationMap) const
+{
+
+  // Create temporary map so can delete elements whilst still iterating over them
+  ClusterAssociationMap tempMap(clusterAssociationMap);
+
+  for(const Cluster *const pCluster : clusterVector) 
+  {
+      const ClusterAssociationMap::const_iterator primaryMapIter = clusterAssociationMap.find(pCluster);
+
+      if(primaryMapIter == clusterAssociationMap.end())
+          continue;
+
+      const ClusterSet &primaryForwardAssociations(primaryMapIter->second.m_forwardAssociations);
+
+      // loop through the primary associations of each cluster
+      // remove clusters that are present in secondary associations of other primary clusters 
```

Comment:
I typically try to keep comments to one line, where something needs to be explained and/or isn't obvious from the code, trying to make the variable and type names as self-describing as possible.